### ISSUE TITLE
Consider moving `UglifyJsPlugin` out of main package.

### ISSUE THREAD

**Issue by wtgtybhertgeghgtwtg** [webpack/webpack#3673]
Relative to other plugins, it has a disproportionate number of dependencies.  Additionally, with the advent of [`babili`](https://github.com/babel/babili), `uglify` is no longer a necessary part of a production build process.

**Comment by bebraw** [comment]
Yeah, agreed. It's also pretty hard to use the harmony branch of Uglify due to this. Basically you have to fork webpack and then tweak `package.json` to point to it.

This is probably a webpack 3 change, though, as it's too late to so the change for 2 if the others agree.

**Comment by mikeerickson** [comment]
@wtgtybhertgeghgtwtg @bebraw What would it take to replace `UglifyJs` with `Babili` so that we can get clean (error free) results when perform `webpack -p`?   Would there need to be a supporting `babili` plugin (ala the UglifyJs plugin) or does it already exist?

**Comment by wtgtybhertgeghgtwtg** [comment]
Add the preset to your `.babelrc`.

**Comment by mikeerickson** [comment]
@wtgtybhertgeghgtwtg are you saying add the `babili` preset to `.babelrc`

**Comment by wtgtybhertgeghgtwtg** [comment]
That's what the `babili` [docs say](https://github.com/babel/babili#usage-1)
```
{
  "presets": ["es2015"],
  "env": {
    "production": {
      "presets": ["babili"]
    }
  }
}
```

**Comment by mikeerickson** [comment]
@wtgtybhertgeghgtwtg makes plenty of sense...

However, how to `disable` use of `UglifyJs` within webpack 2 as to fix this original issue?

---

```
ERROR in bundle.js from UglifyJs
SyntaxError: Unexpected token: operator (>) [./~/ansi-styles/index.js:4,0]
```

**Comment by mikeerickson** [comment]
@wtgtybhertgeghgtwtg FWIW, I have found a rather simple solution

https://www.npmjs.com/package/babili-webpack-plugin

**Comment by jpollard-cs** [comment]
+ for babili - also for anyone wondering why UglifyJS is running in the first place (if you don't have it in your `webpack.config.js` file), check your scripts in your `package.json` - you are most likely using `-p`

**Comment by bebraw** [comment]
FWIW, there's [webpack-contrib/uglifjs-webpack-plugin](https://github.com/webpack-contrib/uglifyjs-webpack-plugin) now. I haven't released it to the public as it's more in an evaluation mode right now so people can have a peek to see if this is a good direction or not.

**Comment by jpollard-cs** [comment]
Thanks @bebraw I'm a little confused though as this seems to have the same Uglify version dependency as Webpack 2? I'm also entirely unclear which version of Uglify on npm correlates with the Uglify harmony branch.

**Comment by bebraw** [comment]
@jpollard-cs Yeah, my current thinking is that the package should treat Uglify as a peer dependency. Then you can decide whether to use the stable version or something else without having to fork anything.

**Comment by johnnyshields** [comment]
Any movement on this?

**Comment by bebraw** [comment]
@johnnyshields No movement. The first step would be to do a PR that makes webpack point at [uglifyjs-webpack-plugin](https://github.com/webpack-contrib/uglifyjs-webpack-plugin) and then later make the breaking change in the next major version.

**Comment by cameron-martin** [comment]
If uglify is removed as a dependency of webpack, then how will minification "just work" when specifying mode as production?

**Comment by prashantverma123** [comment]
Hi, I would like to take this issue. Is it still available? thanks.

**Comment by abhinavsingi** [comment]
Webpack-5 is removing UglifyJsPlugin and using terser-webpack-plugin. So I guess this won't be needed?

**Comment by ShootPassSlam** [comment]
As mentioned above, UglifyJsPlugin has been removed from the dependencies for terser-webpack-plugin. Should this issue be closed?

**Comment by evilebottnawi** [comment]
Yes, we can close issue, also you can using own plugin for minification, terser by default for better DX

**Comment by wtgtybhertgeghgtwtg** [comment]
No, but it might be more accurate to rename it.  All that needs to be done to accomplish what this issue set out to do (externalize minification to potentially red) would be to remove `terser-webpack-plugin` from the default configuration.

**Comment by evilebottnawi** [comment]
@wtgtybhertgeghgtwtg why it is normal and expected:
1. You should have minimizer out of box - young developers have problems with configure webpack - it is very bad, we want better DX and less configurations
2. It is dev dependencies
3. It is part of `0CJS`
4. When you use bundler you want write code so we did it for you

**Comment by wtgtybhertgeghgtwtg** [comment]
`webpack` configuration is difficult, yes, but adding `terser-webpack-plugin` as `webpack` does by default is pretty straightforward.
```js
// webpack.config.js
const TerserPlugin = require('terser-webpack-plugin');

const devtool = 'whatever';
module.exports = {
  devtool,
  optimization: {
    minimizer: [
      new TerserPlugin({
        cache: true,
	parallel: true,
        // Or just true or false, but copying from `WebpackOptionsDefaulter`
        sourceMap: devtool && /source-?map/.test(devtool)
      }),
    ],
  },
};
```
Moreover, I'd argue that a new developer probably wouldn't even have to do that; while I can't know how many projects start from a boilerplate, I do know that a significant and growing portion of `webpack` downloads are transitive from packages that manage `webpack` configuration (`react-scripts`, `next`, `nuxt`, et cetera).
While being a dev dependency does mean it won't inflate a consumer-facing bundle or increase attack vendors, it still takes space and time (the install size of `terser-webpack-plugin` is about 8 MB, most of which doesn't dedupe with the rest of `webpack`) and can be blocked by company proxies on security grounds.

**Comment by evilebottnawi** [comment]
> can be blocked by company proxies on security grounds.

Weird, don't work on company what can block internet

>  I do know that a significant and growing portion of webpack downloads are transitive from packages that manage webpack configuration (react-scripts, next, nuxt, et cetera).

All of them use `terser` because it is one stable solution for minification js right now

It is dev dependencies so you can relax about this, also future `pnp` technologies solve local size problem with node_modules

DX > extra dep in dev deps

**Comment by wtgtybhertgeghgtwtg** [comment]
Many companies use private package registries (i.e. Nexus), which can, among other things, enforce licensing and security policies.
While they do happen to all use `terser`, they do so explicitly (`terser-webpack-plugin` for `react-scripts` and `nuxt`, custom plugin wrapping `terser` for `next`), completely eschewing the default and potentially the version of `terser-webpack-plugin` that `webpack` brings in.  Aside from that, there's no guarantee that `terser` will remain the sole stable solution; this issue was raised back when `webpack` was using `uglify`, after all.

**Comment by evilebottnawi** [comment]
> enforce licensing and security policies.

Using private package registries is not solve problem. Code still have security problems and can have license problems (copy/paste code, using git submodules, invalid licenses and etc). You need more powerful tools for this.

> Aside from that, there's no guarantee that terser will remain the sole stable solution

Yes, but right now it is stable. It is open source, no guaranties (MIT license). 

Here is PR to solve DX and you problems https://github.com/webpack/webpack/pull/6395. But it is not high priority because we have many other problems. You can continue development and in future you can use like:
```
npm i webpack webpack-default-preset
```

If you don't need `terser` just don't install `webpack-defaults-preset`. 

Right now we can't remove `terser` due reasons above.

**Comment by wtgtybhertgeghgtwtg** [comment]
I never said it solved the problem, it's just what some companies do, with the side-effect being development can be halted if a transitive dependency is flagged.  This's happened before with `sockjs` in `webpack-dev-server`.
The point is that including it as a default increases the dependency tree by no small amount, while also being completely unused in many setups.  It doesn't matter if you're using `terser` or `uglify` or `babel-minify`, `react-scripts` or `next` or `nuxt` or, in the future, even `webpack` presets; unless you are using the default, that entire chunk of the dependency tree is unused.

**Comment by evilebottnawi** [comment]
It is just dev deps, what is problem with extra package? 
Disk space? `pnp` solve this
Extra package? In any big projects you already have many dev deps (90% of this you don't use)
Perf? No

Please provide link on **real** problem where developers have problem with this extra dep. Answer - 'i just want' is not enough.

**Comment by wtgtybhertgeghgtwtg** [comment]
A larger dependency tree means more to download, which, besides just needing more disk space, also means more network usage and install time, both of which [negatively impact downstream users](https://github.com/facebook/create-react-app/issues/3880).  This is especially noticeable in containerized environments where it does not install from a cache.
The bottom line is that dependencies are not free.  More than network or disk, they add complexity.  They increase the size of the system, adding more variables, and that means more things that can go wrong, in any number of ways.  While those variables cannot be controlled here, what can be controlled is what gets included in this system.  I am not saying to do something because I just it it, but I am saying that, if a good chunk of those variables, and thus potential issues, can be removed at the cost of a default that is unused by a significant and increasing portion of users, I believe it should be considered.

**Comment by evilebottnawi** [comment]
This problem out of scope webpack, feel free to open issue in `npm`/`yarn`. Here doesn't exists right solution, some uses caches, some `pnp`, some move deps to own container and etc. It is problem related to **any** javascript (and not only) ecosystem.

**Comment by evilebottnawi** [comment]
At the moment we are leaving everything as it is and in the near future we don't change that, because there are more important problems than this, we may revise it in the future and reopen, but not now and not in near future. Thanks

**Comment by evilebottnawi** [comment]
Also removing `terser` doesn't solve https://github.com/facebook/create-react-app/issues/3880, because CRA use terser

**Comment by evilebottnawi** [comment]
Also a lot of packages don't use `files` field n package.json and provide many files what is never used (tests, examples, extra docs, images and etc). It is problem also related to javascript ecosystem. It can be solved only package manager level (warnings or fail on publish/reduce size of register level/other solutions).

**Comment by wtgtybhertgeghgtwtg** [comment]
> It can be solved only package manager level

It's not a problem that can be solved, only alleviated, and only by the developers of the individual packages.  `yarn` and `npm` can't make you make your package simple.  If you include tests, if you include docs, if you include a CLI, if you include defaults, if you include [ASCII pictures of Guy Fieri](https://github.com/babel/babel/blob/f36d07d30334f86412a9d2771880cb566a82a9b6/packages/babel-core/src/api/node.js#L35-L219), that's all entirely your choice, and you may choose to do so for whatever reason you wish.  I may advise against it, I may open issues against it, I may argue my point, I may offer to make a PR, but that's all I can do to alleviate complexity.
